Add an OpenMP multi-threaded parallelization

ymbyun opened the pull request:
I added a couple of features:

- OpenMP parallelization
- OpenMP optimization using first touch (i.e. using memory affinity)

The above two features are enabled by two flags in Makefile:

- -qopenmp (Intel Fortran compiler)
- -DENABLE_OPENMP_AFFINITY

There are a couple of things to note:

- Memory affinity should be used together with process affinity (i.e. CPU binding).
- Currently, MPI (for RI) and OpenMP (for no-RI) can't be used together (i.e. users need at least two executable files).

Please let me know if you have any questions.

bruneval commented:
Dear Young-Moo,
Thank you for your work.
I merged your pull request into a new branch named "ymbyun-add-openmp".
The merge went smoothly with very few conflicts. But there are a few things that need clarification and decision before merging everything in the master.
1. In your email, you said that all your developments were only activated when compiling with -DENABLE_YMBYUN. However there is only one occurrence of  "ENABLE_YMBYUN" in the source code.
2. In your email, you said the "nsize" variable was converted to long integer. However I don't find this in the source code
3. The OMP affinity concept is interesting but it makes the code very difficult to read. Have you quantified the gain of this coding? In my opinion, it needs to be substantial to compensate for the complexity it implies. Do you have tests for this?
4. OMP_STACKSIZE: Could you please tell me for which array it is important to make it larger than the default? In my part (with RI), I need it large in the routine setup_exchange_ri .
5. In the AO -> MO transform (m_eri_ao_mo) I would like you to check out and test how I introduced BLAS routines in replacement for your OMP statemens in the branch . Does it help?

Remarks:
A. There are now a lot of OMP statement in the RI part of the code as well. You don't need to maintain two executables.
B. In OMP, you don't need to set the loop indexes as PRIVATE. This is the default.
C. Could you pull my new branch "ymbyun-add-openmp" into yours and see if everything is still ok.
D. There is a test suite in folder tests. Please run ./run_testsuite to check that everything is ok.

Let me know your answers.
Best

ymbyun commented:
Hi Fabien,

Thank you for your reply. Glad to hear that the merge went smoothly. Below are my answers to your questions:

1. It's because I chose only three among all my changes and broke them down into three branches. If you uncompress the molgw-1.F-Young-Moo-2018.11.01.tar.gz file that I sent you a week ago and run "grep ENABLE_YMBYUN *.f90" in the src directory, you will get a lot of results. Without the ENABLE_YMBYUN flag, my version and your version give the same output.
2. It's in another branch (i.e. Change-int4-to-int8).
3. It's in another branch (i.e. Add-spectral-function-method).
4. That's the point! The affinity thing can lead to ~10% of speedup in certain circumstances (e.g. mainly AMD CPUs), but it has a few drawbacks: (i) it makes a code longer and thus harder to read, (ii) it gives two executable files, (iii) it requires all cores on a node to be used, and (iv) it has no effect on Intel CPUs. I'll get affinity test results soon, and I'll send them to you as soon as I get them.
5. It's the integrals(:,:,:,:) array in the calculate_eri_4center() subroutine. You can find my comment there.
6. I'll test it. Please note that we can turn off Intel's OpenMP-BLAS routines at run time by using "call mkl_set_num_threads(1)". I expect that using my OpenMP will give better performance than using OpenMP-BLAS, because it reduces thread creation/destruction overheads.

Below are my comments for your remarks:

1. I didn't know that, because I didn't read the 1.H source code yet. Glad to know that I don't need to compile twice.
2. It's for the balance between a long code and a safe code. I have always used DEFAULT(NONE) in my OpenMP code for safety, but this time, I didn't, because it makes a code too long. It's up to you whether to set loop indices to PRIVATE or not.
3. Yes, I can do that.
4. I'll do it.

Hope I answered your questions clearly. Please let me know if you have any more questions. I'll send you my test results soon.

Best,
Young-Moo

P.S. You can find my Makefile for OpenMP in my first pull request (i.e. Change-int4-to-int8).

ymbyun commented:
Hi Fabien,

Here are some test results:

- ymbyun-add-openmp: I ran run_testsuite.py and got 1 FAIL (GW HOMO energy of h2_pbe_gw.in) and 74 (= 521 - 446 - 1) NOT FOUNDs. I don't know why. I used MKL 2015, LIBINT 2.2.0, and LIBXC 3.0.0.
- m_eri_ao_mo.f90: Fortran DOT_PRODUCT is replaced by BLAS DGEMM, but we should be careful about it because of a couple of reasons. First, it gives up portability between Intel and GNU compilers. MOLGW compiled by gfortran will be very slow because DGEMM is not parallelized by gfortran. Second, it's not a good idea to use OpenMP-DGEMM and OpenMP-MATMUL inside loops, because it increases thread creation/destruction overheads by the number of iterations. In order to minimize the overheads, OpenMP (and MPI as well) should parallelize the outermost loop whenever possible. Could you do careful speedup tests for ifort vs gfortan and DOT_PRODUCT vs DGEMM?
- Affinity: As I told you before, affinity is mainly for AMD CPUs. Affinity is like overclocking CPUs or GPUs; gain from it is ~10%. Affinity gave me ~10% of speedup for AMD Opteron CPUs. I'm testing Intel Xeon CPUs now. If you can access AMD Ryzen/Threadripper/Epyc CPUs, could you test them using my version of 1.F (you don't need to modify the source code)? I expect that affinity will give the highest gain for 2nd-Gen Threadripper because it has only 2 memory controllers for 4 dies.

Please let me know if you have any questions.

Best,
Young-Moo

bruneval commented:
Hi Young-Moo

1. I'm currently investigating the FAILED test. I think I introduced a bug.
The 'NOT FOUND' error comes from tests that absolutely require SCALAPACK. I should find a way to switch them off cleanly when your compilation does not have Scalapack. This is in my to-do list.
2. I've to test this. But in my experience in the other places of the code, the gain using the threaded DGEMM is spectacular even though the code has to open and close a parallel section at each call.
Let me test this.
3. Concerning affinity, 10 % speed-up on 1 type of CPU is not much... I don't have access to AMD processors. Maybe we should focus on the very section where it matters. For instance, you implemented the affinity for "find_qp_energy_linearization". And I can bet that this does not change the overall timing.
Naively, I would say that this implementation makes sense in the AO->MO transform and in the setup_exchange. Would you agree with this?

bruneval commented:
Also please explain to me what you do in linear.f90 , subroutine chi_to_vchiv .
I don't understand the purpose of the array spstst2in(mpspin,pstate,mstate)
Thanks.

ymbyun commented:
Hi Fabien,

I'm lucky to get your message before going to bed. Below are my answers to your questions:

1. Glad to hear that I didn't mess up with your code. Please let me know if you can't find the origin of the FAILED test result. Maybe I caused the error. (But, please note that without the ENABLE_YMBYUN flag, my changes have no effect on your code.)
2. Glad to learn something new. Please send me the test results for DOT_PRODUCT/DGEMM when they're available. Also, please note that we may need to make a DOT_PRODUCT/DGEMM flag for gfortran users. Without parallelized m_eri_ao_mo.f90, it will take forever to build W.
3. Don't worry too much about affinity. I implemented affinity everywhere not for speedup, but for consistency. You can leave affinity if it's at hotspots (i.e. bottlenecks), but you can remove it if it's not. Affinity is a small thing. I'll handle it.
4. Without the array spstst2in(mpspin,pstate,mstate), loops depend on each other and thus we get random numbers. I left some comments about the dependency issue there. The best way to understand the issue is to remove spstst2in(mpspin,pstate,mstate) and see what happens.

Please let me know if you have any more questions.

Thank you,
Young-Moo

bruneval commented:
Hi Young-Moo,
Good that you got the previous message before going to bed. You'll have this one when waking up.
The failed test was on your side finally. But it is fixed now.
When **not** zeroing the eri_4center array in linear_response.f90, you changed the behavior of the subroutine.
The "not found" tests were my mistake. They are fixed by the last commit.

Could you please perform a pull request for the other 2 developments you implemented?
Thanks.

Fabien

bruneval commented:
I performed a benchmark on the new implementation with or without affinity (last commit in my branch).
It looks like it helps a bit on Intel Skylake (48 cores / node). However in my benchmark, 95% of the time is consumed in the pre-calculation of the 4-center integrals. And there, the affinity coding does not change anything.
Do you draw the same conclusion on your tests?

ymbyun commented:
Hi Fabien,

I think it's time for you to go to bed. Below are my comments and answers:

- Oops! Sorry about the bug! I tried to give you bug-free changes, but I made one. Thank you for letting me know where the bug is. I have just fixed it in my version of 1.F, which I'm using for my current project. It seems like the bug affects only a small molecule with small basis.
- I already sent you two other changes a couple of weeks ago. They are Pull Requests #3 and #4.

There are some caveats about affinity:

- Affinity should affect old and new AMD CPUs (all of them are MCM) and new Intel CPUs (Nehalem/Sandy Bridge/Ivy Bridge/Haswell/Broadwell/Sky Lake are monolithic, but Cascade Lake, released a couple of weeks ago, is MCM). Thus, affinity should have little effect on Sky Lake.
- Manuals say that OMP_PROC_BIND and OMP_PLACES are equivalent to KMP_AFFINITY, but I found that it is not the case. Only KMP_AFFINITY works well for MKL codes on Intel CPUs. Please modify the PBS example script that I sent you two weeks ago.
- The more memory is used, the stronger NUMA effects are. To test the effect of affinity, 10-100GB of memory should be used. To use 10-100GB of memory (i.e. to overcome the 4GB array limit), my another change (i.e. change-int4-to-int8) should be merged first. If you use a large amount of memory, you will find that pre-SCF/SCF/post-SCF use similar amounts of compute time.
- Affinity is a small thing for computations, but a big thing for politics. I wrote in my proposal for next year's NERSC compute time "We converted a MPI code to an OpenMP one to best utilize NERSC computing resources, which are especially designed for shared-memory programming. In addition, we optimized it using memory affinity to enhance memory bandwidth and lower memory latency as NERSC recommended.", which will increase the chance for us to get some compute time from NERSC next year because it sounds cool.

Thank you for your response and test. Please let me know if you have any questions about my changes.

Best,
Young-Moo

bruneval commented:
Hello!
I did not pay attention to the other two pull requests. Thank you for splitting your developments like this.
This is much cleaner.
I have a last question for the OPENMP part. In linear_response.f90 line 1208, there is this comment:
   ! ymbyun 2018/05/22
   ! NOTE: calculate_eri_4center_eigen() is OpenMP parallelized without being orphaned. 
I am not familiar with the concept of "orphan" . Could you please explain this to me?
Thanks

Fabien

ymbyun commented:
Hi Fabien,

You requested me to break down my changes into multiple branches last summer, so I just did it.

In linear_response.f90, PARALLEL is at an inner loop and calculate_eri_4center_eigen() is outside the loop, so calculate_eri_4center_eigen() should contain PARALLEL inside. If we move PARALLEL in linear_response.f90 from the inner loop to the outermost loop to reduce overheads, calculate_eri_4center_eigen() should not contain PARALLEL inside (i.e. it should be orphaned). Here is an example of an orphaned directive: 

https://computing.llnl.gov/tutorials/openMP/samples/Fortran/omp_orphan.f

Other two changes are much shorter than this OpenMP change, so they won't take you long.

Best,
Young-Moo

bruneval commented:
This branch has been merged into the master even though github does not detect it.
I did it manually on my local computer and then pushed the result on github.

ymbyun commented:
Hi Fabien,

Thank you for merging this branch into the master. I have just pulled the master and checked that it works well.

Young-Moo
